File: clientlib/java/testapp/src/TestClient.java
import io.flic.fliclib.javaclient.*;
import io.flic.fliclib.javaclient.enums.*;

import java.io.IOException;

/**
 * Test Client application.
 *
 * This program attempts to connect to all previously verified Flic buttons by this server.
 * Once connected, it prints Down and Up when a button is pressed or released.
 * It also monitors when new buttons are verified and connects to them as well. For example, run this program and at the same time the ScanWizard program.
 */
public class TestClient {

    private static ButtonConnectionChannel.Callbacks buttonCallbacks = new ButtonConnectionChannel.Callbacks() {
        @Override
        public void onCreateConnectionChannelResponse(ButtonConnectionChannel channel, CreateConnectionChannelError createConnectionChannelError, ConnectionStatus connectionStatus) {
            System.out.println("Create response " + channel.getBdaddr() + ": " + createConnectionChannelError + ", " + connectionStatus);
        }

        @Override
        public void onRemoved(ButtonConnectionChannel channel, RemovedReason removedReason) {
            System.out.println("Channel removed for " + channel.getBdaddr() + ": " + removedReason);
        }

        @Override
        public void onConnectionStatusChanged(ButtonConnectionChannel channel, ConnectionStatus connectionStatus, DisconnectReason disconnectReason) {
            System.out.println("New status for " + channel.getBdaddr() + ": " + connectionStatus + (connectionStatus == ConnectionStatus.Disconnected ? ", " + disconnectReason : ""));
        }

        @Override
        public void onButtonUpOrDown(ButtonConnectionChannel channel, ClickType clickType, boolean wasQueued, int timeDiff) throws IOException {
            System.out.println(channel.getBdaddr() + " " + (clickType == ClickType.ButtonUp ? "Up" : "Down"));
        }
    };

    public static void main(String[] args) throws IOException {
        final FlicClient client = new FlicClient("localhost");
        client.getInfo(new GetInfoResponseCallback() {
            @Override
            public void onGetInfoResponse(BluetoothControllerState bluetoothControllerState, Bdaddr myBdAddr,
                                          BdAddrType myBdAddrType, int maxPendingConnections, int maxConcurrentlyConnectedButtons,
                                          int currentPendingConnections, boolean currentlyNoSpaceForNewConnection, Bdaddr[] verifiedButtons) throws IOException {

                for (final Bdaddr bdaddr : verifiedButtons) {
                    client.addConnectionChannel(new ButtonConnectionChannel(bdaddr, buttonCallbacks));
                }
            }
        });
        client.setGeneralCallbacks(new GeneralCallbacks() {
            @Override
            public void onNewVerifiedButton(Bdaddr bdaddr) throws IOException {
                System.out.println("Another client added a new button: " + bdaddr + ". Now connecting to it...");
                client.addConnectionChannel(new ButtonConnectionChannel(bdaddr, buttonCallbacks));
            }
        });
        client.handleEvents();
    }
}

File: clientlib/python/test_client.py
#!/usr/bin/env python3

# Test Client application.
#
# This program attempts to connect to all previously verified Flic buttons by this server.
# Once connected, it prints Down and Up when a button is pressed or released.
# It also monitors when new buttons are verified and connects to them as well. For example, run this program and at the same time the scan_wizard.py program.

import fliclib

client = fliclib.FlicClient("localhost")

def got_button(bd_addr):
	cc = fliclib.ButtonConnectionChannel(bd_addr)
	cc.on_button_up_or_down = \
		lambda channel, click_type, was_queued, time_diff: \
			print(channel.bd_addr + " " + str(click_type))
	cc.on_connection_status_changed = \
		lambda channel, connection_status, disconnect_reason: \
			print(channel.bd_addr + " " + str(connection_status) + (" " + str(disconnect_reason) if connection_status == fliclib.ConnectionStatus.Disconnected else ""))
	client.add_connection_channel(cc)

def got_info(items):
	print(items)
	for bd_addr in items["bd_addr_of_verified_buttons"]:
		got_button(bd_addr)

client.get_info(got_info)

client.on_new_verified_button = got_button

client.handle_events()
